ring.go

```diff
@@ -213,32 +216,66 @@ type ringShards struct {
 
 	mu       sync.RWMutex
 	hash     ConsistentHash
-	shards   map[string]*ringShard // read only
-	list     []*ringShard          // read only
+	shards   map[string]*ringShard // read only, updated by SetAddrs
+	list     []*ringShard          // read only, updated by SetAddrs
 	numShard int
 	closed   bool
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
-	list := make([]*ringShard, 0, len(shards))
+	c := &ringShards{
+		opt: opt,
+	}
+	c.SetAddrs(opt.Addrs)
+
+	return c
+}
+
+// SetAddrs must not be called concurrently.
+func (c *ringShards) SetAddrs(addrs map[string]string) {
+	shards := make(map[string]*ringShard)
+	if c.closed {
+		return
+	}
+
+	for k, shard := range c.shards {
+		if addr, ok := addrs[k]; ok && shard.addr == addr {
+			shards[k] = shard
+		} else {
+			defer func() {
+				err := shard.Client.Close()
+				if err != nil {
+					log.Printf("Failed to close ring shard client %s: %v", shard.addr, err)
```

Comment:
I think it makes sense to log shard name as well

Reply:
done